File: vault-ios/VaultCommon/Features/FileIcon/FileIconCache.swift
import UIKit
import VaultMobile

public class FileIconCache {
    private let mobileVault: MobileVault

    private var cache: [FileIconCacheKey: UIImage] = [:]

    public init(mobileVault: MobileVault) {
        self.mobileVault = mobileVault
    }

    public func getIcon(props: FileIconProps, scale: Int) -> UIImage {
        let key = FileIconCacheKey(props: props, scale: scale)

        if let icon = cache[key] {
            return icon
        } else {
            let icon = buildIcon(props: props, scale: scale)
            cache[key] = icon
            return icon
        }
    }

    private func buildIcon(props: FileIconProps, scale: Int) -> UIImage {
        let png = mobileVault.fileIconPng(props: props, scale: UInt32(scale))

        return UIImage(data: Data(png.png), scale: Double(scale))!
    }
}

struct FileIconCacheKey: Hashable {
    let props: FileIconProps
    let scale: Int
}

File: vault-android/app/src/main/java/net/koofr/vault/features/fileicon/FileIconCache.kt
package net.koofr.vault.features.fileicon

import android.graphics.BitmapFactory
import androidx.compose.ui.graphics.ImageBitmap
import androidx.compose.ui.graphics.asImageBitmap
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.components.ActivityRetainedComponent
import dagger.hilt.android.scopes.ActivityRetainedScoped
import net.koofr.vault.FileIconProps
import net.koofr.vault.MobileVault
import javax.inject.Inject

data class FileIconCacheKey(val props: FileIconProps, val scale: Int)

class FileIconCache @Inject constructor(val mobileVault: MobileVault) {
    private val cache: MutableMap<FileIconCacheKey, ImageBitmap> = mutableMapOf()

    @Synchronized
    fun getIcon(props: FileIconProps, scale: Int = 3): ImageBitmap {
        val key = FileIconCacheKey(props, scale)

        val cachedIcon = cache[key]

        return if (cachedIcon != null) {
            cachedIcon
        } else {
            val icon = buildIcon(props, scale)
            cache[key] = icon
            icon
        }
    }

    private fun buildIcon(props: FileIconProps, scale: Int): ImageBitmap {
        val png = mobileVault.fileIconPng(props = props, scale = scale.toUInt())

        val pngByteArray = png.png

        val bitmap = BitmapFactory.decodeByteArray(pngByteArray, 0, pngByteArray.size)

        return bitmap.asImageBitmap()
    }
}

@Module
@InstallIn(ActivityRetainedComponent::class)
object FileIconCacheModule {
    @ActivityRetainedScoped
    @Provides
    fun provideFileIconCache(mobileVault: MobileVault): FileIconCache {
        return FileIconCache(mobileVault)
    }
}
